Use the new collector internal service for process information

Molter73 opened the pull request:
## Description

This PR makes it so collector starts using the new unified internal service for communicating with sensor. For now the implementation only handles process information, but lays the groundwork to also move network flows into this new internal service.

In order to better abstract how we forward messages, a new `CollectorOutput` class has been added. The responsibility for this class is to hold all gRPC clients that we might need for sending messages to collector, both through the new internal service and the legacy ones, as well as ensuring the connection between sensor and collector is up. This change means the responsibility for the `ProcessSignalClient` class has been reduced to simply sending the message through the grpc client it holds. A new client has also been added for the new service.

Alongside this new `CollectorOutput` class, a new formatter has been added to generate messages that will be sent through the new service and the existing process signal handler has been modified to start using it.

Sibling PR stackrox/stackrox#14652

## Checklist
- [x] Investigated and inspected CI test results
- [x] Updated documentation accordingly

**Automated testing**
  - [x] Added unit tests
  - [ ] Added integration tests
  - [ ] Added regression tests

If any of these don't apply, please comment below.

## Testing Performed

Deployed this version of collector with the sensor version from stackrox/stackrox#14652 and checked process information is still available in the Risk tab.
Modified the existing integration tests to use the new service.

Molter73 committed:
Basic implementation using the new collector internal service

Molter73 committed:
Add CollectorOutput component

Molter73 committed:
Cleanup

Molter73 committed:
Update mock sensor to use the newer collector-sensor service

Molter73 committed:
Restore ProcessSignalFormatter, add SensorClientFormatter

This change makes it possible to have backwards compatibility with older
sensor versions that might no implement the new service.

Molter73 committed:
Unify client reconnection thread in CollectorOutput

Molter73 committed:
Remove unneeded operator= overloads

Molter73 committed:
Add some unit tests

Molter73 committed:
Use an env variable for legacy services configuration

Molter73 committed:
Restore some log messages needed for testing in the main repo

Molter73 commented on collector/lib/SensorClientFormatter.cpp:
This file is essentially a copy of `ProcessSignalFormatter.cpp`, but building a `sensor::MsgFromCollector` instead of a `storage::ProcessSignal`.

Molter73 commented on collector/lib/SensorClientFormatter.h:
This file is essentially a copy of `ProcessSignalFormatter.h`, but building a `sensor::MsgFromCollector` instead of a `storage::ProcessSignal`.

Molter73 commented on collector/lib/CollectorOutput.cpp:
```diff
@@ -0,0 +1,144 @@
+CollectorOutput::CollectorOutput(const CollectorConfig& config) {
+  use_sensor_client_ = !config.UseLegacyServices();
+
+  if (config.grpc_channel != nullptr) {
```
We had it like that, with the current setup you could have both a grpc output and a stdout output, which I think could be useful for debugging. I'm happy to roll it back if it is confusing or if you feel like we should have a deeper conversation on it.

Stringy replied:
I agree that we want both, but what I meant specifically is that we probably want to have the stdout output if we don't have the grpc channel, rather than relying on the config

```
if (grpc) {
    // init grpc bits
}

if (!grpc || use_stdout) {
   // init stdout bits
}
```

Molter73 replied:
I honestly don't quite like this approach, but given it was the default before, I'll roll the behavior back.

Molter73 commented on collector/lib/SignalServiceClient.h:
```diff
@@ -23,53 +23,56 @@ class ISignalServiceClient {
   std::unique_ptr<IDuplexClientWriter<SignalStreamMessage>> writer_;
 
-  bool first_write_;
+  bool first_write_{};
```
I think I had a linter error at some point, leaving it as it was is not a guarantee it will be initialized in constructors that don't explicitly initialize the variable.

Molter73 replied:
Confirmed, it's a linter error that I was fixing with this
![image](https://github.com/user-attachments/assets/a47292f5-8181-4eea-852d-f27295f381e6)

Molter73 commented on collector/lib/SensorClientFormatter.h:
```diff
@@ -0,0 +1,70 @@
+
+namespace collector {
+
+class SensorClientFormatter : public ProtoSignalFormatter<sensor::MsgFromCollector> {
```
Yeah, I forgot I wanted to add docs for all the new stuff, I'll get it sorted out.

Stringy commented on collector/lib/ProcessSignalFormatter.cpp:
```diff
@@ -65,7 +65,7 @@ ProcessSignalFormatter::ProcessSignalFormatter(
 }
 
-ProcessSignalFormatter::~ProcessSignalFormatter() {}
+ProcessSignalFormatter::~ProcessSignalFormatter() = default;
```
I'm not sure I understand - the class and this destructor is declared in the header so we should just be able to do `~ProcessSignalFormatter() = delete;` there, no? Am I missing some weird C++ism?

Molter73 replied:
Ok, I tried it to be fully certain about why this is not possible. When I moved the destructor definition to the header file, I got this error:
```
[ 80%] Building CXX object collector/lib/CMakeFiles/collector_lib.dir/ProcessSignalHandler.cpp.o
In file included from /usr/include/c++/11/memory:76,
                 from /usr/local/include/grpcpp/channel.h:22,
                 from /home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/ProcessSignalHandler.h:4,
                 from /home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/ProcessSignalHandler.cpp:1:
/usr/include/c++/11/bits/unique_ptr.h: In instantiation of ‘void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = collector::system_inspector::EventExtractor]’:
/usr/include/c++/11/bits/unique_ptr.h:361:17:   required from ‘std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = collector::system_inspector::EventExtractor; _Dp = std::default_delete<collector::system_inspector::EventExtractor>]’
/home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/ProcessSignalFormatter.h:31:3:   required from here
/usr/include/c++/11/bits/unique_ptr.h:83:23: error: invalid application of ‘sizeof’ to incomplete type ‘collector::system_inspector::EventExtractor’
   83 |         static_assert(sizeof(_Tp)>0,
      |                       ^~~~~~~~~~~
gmake[2]: *** [collector/lib/CMakeFiles/collector_lib.dir/build.make:510: collector/lib/CMakeFiles/collector_lib.dir/ProcessSignalHandler.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:627: collector/lib/CMakeFiles/collector_lib.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

Now, what that nice error is trying to say is that we are trying to call the destructor for this member in the class:
```cpp
  std::unique_ptr<system_inspector::EventExtractor> event_extractor_;
```

But, because we are trying to limit the places where we include the falco headers, in `ProcessSignalFormatter.h` the `system_inspector::EventExtractor` class is forward declared like this:
```cpp
namespace collector::system_inspector {
class EventExtractor;
}
```

That means the header doesn't know what the destructor for `EventExtractor` looks like and therefore is unable to infer the default destructor for the whole formatter. In `ProcessSignalFormatter.cpp` we include the whole header file for the extractor like this:
```cpp
#include "system-inspector/EventExtractor.h"
```

And so, `ProcessSignalFormatter::~ProcessSignalFormatter()` needs to be defined in the source file.

Molter73 replied:
This took me quite a bit of time to figure out when I stumbled upon it a few days ago BTW, in case you couldn't tell from the explanation.

Stringy replied:
C++ is definitely one of the languages of all time. 

Makes sense, thanks for the thorough explanation!

Molter73 commented on collector/lib/SensorClientFormatter.cpp:
```diff
@@ -0,0 +1,377 @@
+    // Collection of process lineage information should stop at the container
+    // boundary to avoid collecting host process information.
+    //
+    // In back-ported eBPF probes, `m_vpid` will not be set for containers
```
This was addressing a specific bug on RHEL 7, right? Given the modern probe doesn't run on those systems, I think we can just get rid of it, newer systems never suffered from the issue IIRC.

Molter73 commented on collector/lib/SensorClientFormatter.cpp:
I considered using some templating to try and reduce duplication, however, the plan is to remove the old services during our next release cycle, so I don't see much value in complicating this with templates only to remove them in a few months.

Molter73 committed:
Address PR comments

- Rollback no grpc channel behaviour.
- Add documentation to new classes.
- Remove unneeded workaround for old kernels.
- Minor refactorings in SensorClientFormatter.

Molter73 committed:
Minor cleanup

JoukoVirtanen commented on collector/lib/SensorClient.cpp:
```diff
@@ -0,0 +1,47 @@
+
+  if (first_write_ && msg.has_process_signal()) {
+    first_write_ = false;
+    return SignalHandler::NEEDS_REFRESH;
```
It seems that `SignalHandler::NEEDS_REFRESH` is returned when `first_write_` is true and that is the case only after `Refresh` has been called. So, `SignalHandler::NEEDS_REFRESH` is returned after a refresh has been done and returning `SignalHandler::NEEDS_REFRESH` does not result in a refresh, as the name implies. It does result in `SendMsg` being called again. In which case `first_write_` will be true and this will be skipped. I don't understand what returning here instead of sending the message accomplishes.

Molter73 replied:
This is poor naming conventions on my part. `NEEDS_REFRESH` is used when the process stream has just started and it causes collector to send all processes that are stored in the threadinfo table, at startup this sends all scraped processes, while running it will ensure any processes that were not sent as part of a disconnect to be sent on reconnection.

I probably need to come up with a better term than `Refresh` for the method called after grpc is reconnected. I'll think it over and rename the method.

Molter73 commented on collector/lib/SensorClient.h:
```diff
@@ -0,0 +1,91 @@
+  bool Refresh() override { return true; }
+
+  SignalHandler::Result SendMsg(const sensor::MsgFromCollector& msg) override {
+    LogProtobufMessage(msg);
```
I can add the check in a separate PR. Do keep in mind this is only done when the stdout output is used which is meant for debugging and should not run in production, so even if it is not ideal, it should not have an impact collector.

Molter73 committed:
Rename Refresh to Recreate

JoukoVirtanen commented on collector/lib/SensorClientFormatter.cpp:
```diff
@@ -0,0 +1,352 @@
+using Timestamp = google::protobuf::Timestamp;
+using TimeUtil = google::protobuf::util::TimeUtil;
+
+namespace {
```
Not saying this is wrong, but how did you decide to use an anonymous namespace rather than declaring these as private?

JoukoVirtanen commented:
This needs more explanation of how you validated your change. You should mention your stackrox/stackrox PR and how that passing CI validates the changes here. Since you are requiring that this works with old versions of sensor, you should also validate that. You could create a stackrox/stackrox PR that just updates the sensor version and sets the environment variable that you introduced.

Other than that, this looks good.

JoukoVirtanen reviewed (approved):
LGTM!

Molter73 committed:
Use pragma once on new header files
